proto/docs/p4runtime.md

```diff
@@ -515,3 +520,176 @@ messages refer to singleton ports by default.
 
 *TODO: Discuss potential use of translation, resolution and watch annotations to
 support cases like multicast to LAGs or clone to arbitrary destination.*
+
+## PSA Metadata Translation
+
+The *Portable Standard Architecture* (PSA) defines intrinsic metadata, whose
+types are undefined, since they may be different on different PSA targets.
+In order to have uniform programming of multiple PSA
+targets, a centralized SDN controller may define its own types and numbering of
+such PSA intrinsic metadata. In particular, different PSA devices may have
+different types and numbering schemes for ports and class of service. For such
+metadata, a translation between the controller's metadata values and the
+corresponding target specific metadata values is required at runtime.
+
+![Controllers with P4Runtime channels to multiple PSA devices](sdn_controller.svg)
+
+The above figure illustrates a motivating example, where an SDN controller is
+controlling multiple P4 targets in a fabric. Switches 1, 2, 3 etc.
+are all targets with different PSA devices, each defining its own port and
+class of service types. In order to uniformly program the fabric, the
+SDN controller may define 32 bit type for ports and 8 bit type for class of
+service. The controller may also define a common number space for ports and
+cannonical classes of service across the fabric. The PSA devices may define
+much shorter bitwidths for these metadata in order to save precious on-chip
+memory in the switch silicon. Furthermore, the values of ports and classes of
+service in the dataplane are PSA device-specific and may be different from the
+values used in the controller(s).
+
+In order to support such SDN use case, P4Runtime requires translation of port
+and class of service metadata values between the controller's space and the
+PSA device's space as needed. Such translation is enabled by a standard
+P4Runtime annotation `@p4runtime_translation`. The annotation value is a
+string identifying the metadata type to be translated. Specifically, we define
+* `@p4runtime_translation("port")` for port translation.
+* `@p4runtime_translation("cos")` for class of service translation.
+
+The controller's port/cos bitwidth, the valid SDN port/cos numbers, and their
+mapping to device specific port/cos numbers should be captured in the switch
+config. The P4Runtime service on the target is expected to use the mapping defined
```

Comment:
I may be asking a brain-dead question here, but just trying to figure out the intended scenarios you are trying to cover with this, or not cover.

Are you expecting that some controllers will be implemented with a 32-bit port number "compiled into its code", and such controller processes would be _unable_ to communicate with P4Runtime service devices that had a 64-bit port number in their switch config, until and unless one or both sides are changed so that their port bit widths agree with each other?

Reply:
I am assuming that all controllers and switches for an SDN domain will agree on a SDN port id type, and use the number space defined by that type to allocate SDN numbers to all ports. This type does not need to be part of the P4 program, because it is already defined in the switch config.